pkg/hooks/manager.go

```diff
@@ -252,83 +257,165 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
 			return nil, err
 		}
 		domainSpecXML = result.GetDomainXML()
+	case hooksV1alpha3.Version:
+		client := hooksV1alpha3.NewCallbacksClient(conn)
+		result, err := client.OnDefineDomain(ctx, &hooksV1alpha3.OnDefineDomainParams{
+			DomainXML: domainSpecXML,
+			Vmi:       vmiJSON,
+		})
+		if err != nil {
+			log.Log.Reason(err).Error("Failed to call OnDefineDomain")
+			return nil, err
+		}
+		domainSpecXML = result.GetDomainXML()
 	default:
 		log.Log.Errorf("Unsupported callback version: %s", callback.Version)
 	}
 
 	return domainSpecXML, nil
 }
 
+func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cloudinit.CloudInitData) ([]byte, []byte, []byte, error) {
+	vmiJSON, err := json.Marshal(vmi)
+	if err != nil {
+		return []byte{}, []byte{}, []byte{}, fmt.Errorf("failed to marshal VMI spec, err: %v", err)
```

Comment:
You can return `nil` instead of `[]byte{}`.

Reply:
```diff
--- a/pkg/hooks/manager.go
+++ b/pkg/hooks/manager.go
@@ -262,7 +262,7 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
 func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cloudinit.CloudInitData) ([]byte, []byte, []byte, error) {
        vmiJSON, err := json.Marshal(vmi)
        if err != nil {
-               return []byte{}, []byte{}, []byte{}, fmt.Errorf("failed to marshal VMI spec, err: %v", err)
+               return nil, nil, nil, fmt.Errorf("failed to marshal VMI spec, err: %v", err)
        }

        // To be backward compatible to sidecar hooks still expecting to receive the cloudinit data as a
@@ -275,12 +275,12 @@ func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cl
        }
        cloudInitNoCloudSourceJSON, err := json.Marshal(cloudInitNoCloudSource)
        if err != nil {
-               return []byte{}, []byte{}, []byte{}, fmt.Errorf("failed to marshal CloudInitNoCloudSource: %v, err: %v", cloudInitNoCloudSource, err)
+               return nil, nil, nil, fmt.Errorf("failed to marshal CloudInitNoCloudSource: %v, err: %v", cloudInitNoCloudSource, err)
        }

        cloudInitDataJSON, err := json.Marshal(cloudInitData)
        if err != nil {
-               return []byte{}, []byte{}, []byte{}, fmt.Errorf("failed to marshal CloudInitData, err: %v", err)
+               return nil, nil, nil, fmt.Errorf("failed to marshal CloudInitData, err: %v", err)
 ```
 
 To be squashed in the first commit